nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/java/org/apache/nifi/controller/state/providers/zookeeper/ITZooKeeperStateProvider.java

```diff
@@ -0,0 +1,280 @@
+                return new MockComponentLogger();
+            }
+        });
+    }
+
+    private ZooKeeperStateProvider createProvider(final Map<PropertyDescriptor, String> properties) throws Exception {
+        final ZooKeeperStateProvider provider = new ZooKeeperStateProvider();
+        initializeProvider(provider, properties);
+        provider.enable();
+        return provider;
+    }
+
+    @After
+    public void clear() throws IOException {
+
+
+
+        try {
+            if (provider != null) {
+                provider.onComponentRemoved(componentId);
+                provider.disable();
+                provider.shutdown();
+                clearDirectories();
+            }
+        } finally {
+            if (zkServer != null) {
+                zkServer.shutdown(true);
+                clearDirectories();
+            }
+        }
+    }
+
+    private static void clearDirectories() {
+        if (tempDir != null) {
+            final List<Path> files = Arrays.asList(
+                    dataDir.resolve("version-2/snapshot.0"),
+                    dataDir.resolve("version-2/log.1"),
+                    dataDir.resolve("version-2"),
+                    dataDir.resolve("myid"),
+                    dataDir,
+                    tempDir
+            );
+
+            files.forEach(p -> {
+                try {
+                    if (p != null) Files.deleteIfExists(p);
+                } catch (final IOException ioe) {}
+            });
+        }
+    }
+
+    @Override
+    protected StateProvider getProvider() {
+        return provider;
+    }
+
+    @Test(timeout = 30000)
+    public void testStateTooLargeExceptionThrownOnSetState() throws InterruptedException {
+        final Map<String, String> state = new HashMap<>();
+        final StringBuilder sb = new StringBuilder();
+
+        // Build a string that is a little less than 64 KB, because that's
+        // the largest value available for DataOutputStream.writeUTF
+        for (int i = 0; i < 6500; i++) {
+            sb.append("0123456789");
+        }
+
+        for (int i = 0; i < 20; i++) {
+            state.put("numbers." + i, sb.toString());
+        }
+
+        while (true) {
+            try {
+                getProvider().setState(state, componentId);
+                Assert.fail("Expected StateTooLargeException");
+            } catch (final StateTooLargeException stle) {
+                // expected behavior.
+                break;
+            } catch (final IOException ioe) {
+                // If we attempt to interact with the server too quickly, we will get a
+                // ZooKeeper ConnectionLoss Exception, which the provider wraps in an IOException.
+                // We will wait 1 second in this case and try again. The test will timeout if this
+                // does not succeeed within 30 seconds.
+                Thread.sleep(1000L);
+            } catch (final Exception e) {
+                e.printStackTrace();
```

Comment:
Recommend removing all uses of printStackTrace() to avoid unexpected console output.